File: 3643-zero-array-transformation-ii/zero-array-transformation-ii.py
class Solution:
    def minZeroArray(self, nums: List[int], queries: List[List[int]]) -> int:
        n, q = len(nums), len(queries)
        cur_sum = k = 0
        diff = [0] * (n+1)

        for idx in range(n):
            while cur_sum + diff[idx] < nums[idx]:
                k += 1
                if k > q:
                    return -1

                left, right, val = queries[k-1]
                
                if right >= idx:
                    diff[max(left, idx)] += val
                    diff[right+1] -= val

            cur_sum += diff[idx]

        return k

File: 3643-zero-array-transformation-ii/zero-array-transformation-ii.cpp
class Solution {
public:
    int minZeroArray(vector<int>& nums, vector<vector<int>>& queries) {
        int n = nums.size(), q = queries.size();
        int curSum = 0, k = 0;
        vector<int> diff(n+1, 0);

        for (int idx = 0; idx < n; idx++) {
            while (curSum + diff[idx] < nums[idx]) {
                k++;
                if (k > q) return -1;

                int left = queries[k-1][0], right = queries[k-1][1], val = queries[k-1][2];

                if (right >= idx) {
                    diff[max(left, idx)] += val;
                    diff[right + 1] -= val;
                }
            }
            curSum += diff[idx];
        }
        return k;
    }
};